From f7f8733fc19a7160669e637c0963578f2a433655 Mon Sep 17 00:00:00 2001
From: allen-bst <70563899+allen-bst@users.noreply.github.com>
Date: Wed, 6 Mar 2024 17:35:04 +0800
Subject: [PATCH 117/224] [A13] Pie - Android11 code sync ,change includes -
 (#130)

-  Changes to restrict launch of Bluestacks specific activites,
by third party apps.

- misc rvc-A13 sync.
---
 .../SettingsProvider/res/values/defaults.xml  |   2 +-
 .../SettingsProvider/res/xml/bookmarks.xml    |   2 +
 .../server/NetworkManagementService.java      | 161 +++++++++---------
 .../net/NetworkPolicyManagerService.java      |   2 +-
 .../android/server/wm/ActivityStarter.java    |  12 ++
 5 files changed, 100 insertions(+), 79 deletions(-)

diff --git a/packages/SettingsProvider/res/values/defaults.xml b/packages/SettingsProvider/res/values/defaults.xml
index 3e1f7bee1fe3..482d015d93d6 100644
--- a/packages/SettingsProvider/res/values/defaults.xml
+++ b/packages/SettingsProvider/res/values/defaults.xml
@@ -42,7 +42,7 @@
     <!-- 0 == off, 3 == on -->
     <integer name="def_location_mode">3</integer>
     <bool name="assisted_gps_enabled">true</bool>
-    <bool name="def_netstats_enabled">true</bool>
+    <bool name="def_netstats_enabled">false</bool>
     <bool name="def_usb_mass_storage_enabled">true</bool>
     <bool name="def_wifi_on">false</bool>
     <!-- 0 == never, 1 == only when plugged in, 2 == always -->
diff --git a/packages/SettingsProvider/res/xml/bookmarks.xml b/packages/SettingsProvider/res/xml/bookmarks.xml
index 454f4562a239..34661618315a 100644
--- a/packages/SettingsProvider/res/xml/bookmarks.xml
+++ b/packages/SettingsProvider/res/xml/bookmarks.xml
@@ -47,10 +47,12 @@
     <bookmark
         category="android.intent.category.APP_CALENDAR"
         shortcut="l" />
+    <!--
     <bookmark
         category="android.intent.category.APP_MAPS"
         shortcut="m" />
     <bookmark
+    -->
         category="android.intent.category.APP_MUSIC"
         shortcut="p" />
     <bookmark
diff --git a/services/core/java/com/android/server/NetworkManagementService.java b/services/core/java/com/android/server/NetworkManagementService.java
index 73d9cc759b10..201e02970b2d 100644
--- a/services/core/java/com/android/server/NetworkManagementService.java
+++ b/services/core/java/com/android/server/NetworkManagementService.java
@@ -133,6 +133,8 @@ public class NetworkManagementService extends INetworkManagementService.Stub {
 
     private static final int MAX_UID_RANGES_PER_COMMAND = 10;
 
+    private static final boolean BST_CHANGES_ENABLED = true;
+
     static final int DAEMON_MSG_MOBILE_CONN_REAL_TIME_INFO = 1;
 
     static final boolean MODIFY_OPERATION_ADD = true;
@@ -439,101 +441,106 @@ public class NetworkManagementService extends INetworkManagementService.Stub {
      */
     private void prepareNativeDaemon() {
 
-        // push any existing quota or UID rules
-        synchronized (mQuotaLock) {
+        if (BST_CHANGES_ENABLED)
+        {
+            setFirewallEnabled(false);
+        } else {
+            // push any existing quota or UID rules
+            synchronized (mQuotaLock) {
 
-            mStrictEnabled = true;
+                mStrictEnabled = true;
 
-            setDataSaverModeEnabled(mDataSaverMode);
+                setDataSaverModeEnabled(mDataSaverMode);
 
-            int size = mActiveQuotas.size();
-            if (size > 0) {
-                if (DBG) Slog.d(TAG, "Pushing " + size + " active quota rules");
-                final HashMap<String, Long> activeQuotas = mActiveQuotas;
-                mActiveQuotas = Maps.newHashMap();
-                for (Map.Entry<String, Long> entry : activeQuotas.entrySet()) {
-                    setInterfaceQuota(entry.getKey(), entry.getValue());
-                }
-            }
-
-            size = mActiveAlerts.size();
-            if (size > 0) {
-                if (DBG) Slog.d(TAG, "Pushing " + size + " active alert rules");
-                final HashMap<String, Long> activeAlerts = mActiveAlerts;
-                mActiveAlerts = Maps.newHashMap();
-                for (Map.Entry<String, Long> entry : activeAlerts.entrySet()) {
-                    setInterfaceAlert(entry.getKey(), entry.getValue());
+                int size = mActiveQuotas.size();
+                if (size > 0) {
+                    if (DBG) Slog.d(TAG, "Pushing " + size + " active quota rules");
+                    final HashMap<String, Long> activeQuotas = mActiveQuotas;
+                    mActiveQuotas = Maps.newHashMap();
+                    for (Map.Entry<String, Long> entry : activeQuotas.entrySet()) {
+                        setInterfaceQuota(entry.getKey(), entry.getValue());
+                    }
                 }
-            }
 
-            SparseBooleanArray uidRejectOnQuota = null;
-            SparseBooleanArray uidAcceptOnQuota = null;
-            synchronized (mRulesLock) {
-                size = mUidRejectOnMetered.size();
+                size = mActiveAlerts.size();
                 if (size > 0) {
-                    if (DBG) Slog.d(TAG, "Pushing " + size + " UIDs to metered denylist rules");
-                    uidRejectOnQuota = mUidRejectOnMetered;
-                    mUidRejectOnMetered = new SparseBooleanArray();
+                    if (DBG) Slog.d(TAG, "Pushing " + size + " active alert rules");
+                    final HashMap<String, Long> activeAlerts = mActiveAlerts;
+                    mActiveAlerts = Maps.newHashMap();
+                    for (Map.Entry<String, Long> entry : activeAlerts.entrySet()) {
+                        setInterfaceAlert(entry.getKey(), entry.getValue());
+                    }
                 }
 
-                size = mUidAllowOnMetered.size();
-                if (size > 0) {
-                    if (DBG) Slog.d(TAG, "Pushing " + size + " UIDs to metered allowlist rules");
-                    uidAcceptOnQuota = mUidAllowOnMetered;
-                    mUidAllowOnMetered = new SparseBooleanArray();
+                SparseBooleanArray uidRejectOnQuota = null;
+                SparseBooleanArray uidAcceptOnQuota = null;
+                synchronized (mRulesLock) {
+                    size = mUidRejectOnMetered.size();
+                    if (size > 0) {
+                        if (DBG) Slog.d(TAG, "Pushing " + size + " UIDs to metered denylist rules");
+                        uidRejectOnQuota = mUidRejectOnMetered;
+                        mUidRejectOnMetered = new SparseBooleanArray();
+                    }
+
+                    size = mUidAllowOnMetered.size();
+                    if (size > 0) {
+                        if (DBG) Slog.d(TAG, "Pushing " + size + " UIDs to metered allowlist rules");
+                        uidAcceptOnQuota = mUidAllowOnMetered;
+                        mUidAllowOnMetered = new SparseBooleanArray();
+                    }
                 }
-            }
-            if (uidRejectOnQuota != null) {
-                for (int i = 0; i < uidRejectOnQuota.size(); i++) {
-                    setUidOnMeteredNetworkDenylist(uidRejectOnQuota.keyAt(i),
-                            uidRejectOnQuota.valueAt(i));
+                if (uidRejectOnQuota != null) {
+                    for (int i = 0; i < uidRejectOnQuota.size(); i++) {
+                        setUidOnMeteredNetworkDenylist(uidRejectOnQuota.keyAt(i),
+                                uidRejectOnQuota.valueAt(i));
+                    }
                 }
-            }
-            if (uidAcceptOnQuota != null) {
-                for (int i = 0; i < uidAcceptOnQuota.size(); i++) {
-                    setUidOnMeteredNetworkAllowlist(uidAcceptOnQuota.keyAt(i),
-                            uidAcceptOnQuota.valueAt(i));
+                if (uidAcceptOnQuota != null) {
+                    for (int i = 0; i < uidAcceptOnQuota.size(); i++) {
+                        setUidOnMeteredNetworkAllowlist(uidAcceptOnQuota.keyAt(i),
+                                uidAcceptOnQuota.valueAt(i));
+                    }
                 }
-            }
 
-            size = mUidCleartextPolicy.size();
-            if (size > 0) {
-                if (DBG) Slog.d(TAG, "Pushing " + size + " active UID cleartext policies");
-                final SparseIntArray local = mUidCleartextPolicy;
-                mUidCleartextPolicy = new SparseIntArray();
-                for (int i = 0; i < local.size(); i++) {
-                    setUidCleartextNetworkPolicy(local.keyAt(i), local.valueAt(i));
+                size = mUidCleartextPolicy.size();
+                if (size > 0) {
+                    if (DBG) Slog.d(TAG, "Pushing " + size + " active UID cleartext policies");
+                    final SparseIntArray local = mUidCleartextPolicy;
+                    mUidCleartextPolicy = new SparseIntArray();
+                    for (int i = 0; i < local.size(); i++) {
+                        setUidCleartextNetworkPolicy(local.keyAt(i), local.valueAt(i));
+                    }
                 }
-            }
-
-            setFirewallEnabled(mFirewallEnabled);
-
-            syncFirewallChainLocked(FIREWALL_CHAIN_NONE, "");
-            syncFirewallChainLocked(FIREWALL_CHAIN_STANDBY, "standby ");
-            syncFirewallChainLocked(FIREWALL_CHAIN_DOZABLE, "dozable ");
-            syncFirewallChainLocked(FIREWALL_CHAIN_POWERSAVE, "powersave ");
-            syncFirewallChainLocked(FIREWALL_CHAIN_RESTRICTED, "restricted ");
-            syncFirewallChainLocked(FIREWALL_CHAIN_LOW_POWER_STANDBY, "low power standby ");
-
-            final int[] chains = {
-                    FIREWALL_CHAIN_STANDBY,
-                    FIREWALL_CHAIN_DOZABLE,
-                    FIREWALL_CHAIN_POWERSAVE,
-                    FIREWALL_CHAIN_RESTRICTED,
-                    FIREWALL_CHAIN_LOW_POWER_STANDBY
-            };
 
-            for (int chain : chains) {
-                if (getFirewallChainState(chain)) {
-                    setFirewallChainEnabled(chain, true);
+                setFirewallEnabled(mFirewallEnabled);
+
+                syncFirewallChainLocked(FIREWALL_CHAIN_NONE, "");
+                syncFirewallChainLocked(FIREWALL_CHAIN_STANDBY, "standby ");
+                syncFirewallChainLocked(FIREWALL_CHAIN_DOZABLE, "dozable ");
+                syncFirewallChainLocked(FIREWALL_CHAIN_POWERSAVE, "powersave ");
+                syncFirewallChainLocked(FIREWALL_CHAIN_RESTRICTED, "restricted ");
+                syncFirewallChainLocked(FIREWALL_CHAIN_LOW_POWER_STANDBY, "low power standby ");
+
+                final int[] chains = {
+                        FIREWALL_CHAIN_STANDBY,
+                        FIREWALL_CHAIN_DOZABLE,
+                        FIREWALL_CHAIN_POWERSAVE,
+                        FIREWALL_CHAIN_RESTRICTED,
+                        FIREWALL_CHAIN_LOW_POWER_STANDBY
+                };
+
+                for (int chain : chains) {
+                    if (getFirewallChainState(chain)) {
+                        setFirewallChainEnabled(chain, true);
+                    }
                 }
             }
-        }
 
 
-        try {
-            getBatteryStats().noteNetworkStatsEnabled();
-        } catch (RemoteException e) {
+            try {
+                getBatteryStats().noteNetworkStatsEnabled();
+            } catch (RemoteException e) {
+            }
         }
 
     }
diff --git a/services/core/java/com/android/server/net/NetworkPolicyManagerService.java b/services/core/java/com/android/server/net/NetworkPolicyManagerService.java
index 13722086e214..306d1cf3b9dd 100644
--- a/services/core/java/com/android/server/net/NetworkPolicyManagerService.java
+++ b/services/core/java/com/android/server/net/NetworkPolicyManagerService.java
@@ -925,6 +925,7 @@ public class NetworkPolicyManagerService extends INetworkPolicyManager.Stub {
         try {
             // Boost thread's priority during system server init
             Process.setThreadPriority(Process.THREAD_PRIORITY_FOREGROUND);
+            mUsageStats = LocalServices.getService(UsageStatsManagerInternal.class);
             if (!isBandwidthControlEnabled()) {
                 Slog.w(TAG, "bandwidth controls disabled, unable to enforce policy");
                 // Call countDown to gracefully return from initService
@@ -932,7 +933,6 @@ public class NetworkPolicyManagerService extends INetworkPolicyManager.Stub {
                 return;
             }
 
-            mUsageStats = LocalServices.getService(UsageStatsManagerInternal.class);
             mAppStandby = LocalServices.getService(AppStandbyInternal.class);
             mActivityManagerInternal = LocalServices.getService(ActivityManagerInternal.class);
 
diff --git a/services/core/java/com/android/server/wm/ActivityStarter.java b/services/core/java/com/android/server/wm/ActivityStarter.java
index 505df08a823a..c45790627ee4 100644
--- a/services/core/java/com/android/server/wm/ActivityStarter.java
+++ b/services/core/java/com/android/server/wm/ActivityStarter.java
@@ -125,6 +125,7 @@ import android.os.UserManager;
 import android.service.voice.IVoiceInteractionSession;
 import android.text.TextUtils;
 import android.util.ArraySet;
+import android.util.BstUtils;
 import android.util.DebugUtils;
 import android.util.Pools.SynchronizedPool;
 import android.util.Slog;
@@ -940,6 +941,17 @@ class ActivityStarter {
                     + "} from uid " + callingUid);
         }
 
+        int uid = Binder.getCallingUid();
+        if (uid >= 10000) {
+            int pid = Binder.getCallingPid();
+            String callingApp = BstUtils.getAppNameFromPid(pid);
+            boolean isAppPrivileged =  BstUtils.bstIsCallingAppPrivileged(uid, callingApp);
+            String pkg = intent.getComponent() != null ? intent.getComponent().getPackageName() : "";
+            if (BstUtils.hideBlueStacksPkg(pkg, isAppPrivileged)) {
+               return ActivityManager.START_CLASS_NOT_FOUND;
+            }
+        }
+
         ActivityRecord sourceRecord = null;
         ActivityRecord resultRecord = null;
         if (resultTo != null) {
